### PR TITLE
Rewrite the C++ Demangler

### PR DESCRIPTION
**Pull request by Thog**
This new Demangler provides support to almost every possible mangled
symbols and should behaves like GNU c++filt.

It works on 98.9% of the sdk's symbols and 99.5%
of Puyo Puyo Tetris's symbols.

### CODE REVIEW

File: Ryujinx.HLE/HOS/Diagnostics/Demangler/Demangler.cs [hunk 1]
```diff
@@ -0,0 +1,3316 @@
...
+            }
+            if (ConsumeIf("V"))
+            {
+                Qualifiers |= CV.Volatile;
+            }
+            if (ConsumeIf("K"))
+            {
+                Qualifiers |= CV.Const;
+            }
+            return Qualifiers;
```
Comment (AcK77) [inline Ryujinx.HLE/HOS/Diagnostics/Demangler/Demangler.cs]: Add a new line before

File: Ryujinx.HLE/HOS/Diagnostics/Demangler/Demangler.cs [hunk 2]
```diff
@@ -0,0 +1,3316 @@
...
+            Reference Res = Reference.None;
+            if (ConsumeIf("O"))
+            {
+                Res = Reference.RValue;
+            }
+            else if (ConsumeIf("R"))
+            {
+                Res = Reference.LValue;
+            }
+            return new SimpleReferenceType(Res, null);
```
Comment (AcK77) [inline Ryujinx.HLE/HOS/Diagnostics/Demangler/Demangler.cs]: Add a new line before

File: Ryujinx.HLE/HOS/Diagnostics/Demangler/Demangler.cs [hunk 3]
```diff
@@ -0,0 +1,3316 @@
...
+        }
+
+        private BaseNode CreateNameNode(BaseNode Prev, BaseNode Name, NameParserContext Context)
```
Comment (AcK77) [inline Ryujinx.HLE/HOS/Diagnostics/Demangler/Demangler.cs]: `Prev` > `Previous` ?

File: Ryujinx.HLE/HOS/Diagnostics/Demangler/Demangler.cs [hunk 4]
```diff
@@ -0,0 +1,3316 @@
...
+                Context.FinishWithTemplateArguments = false;
+            }
+
+            return Res;
+        }
+
+        private int ParsePositiveNumber()
+        {
+            string Part = Mangled.Substring(Pos);
+            int NumberLen;
```
Comment (AcK77) [inline Ryujinx.HLE/HOS/Diagnostics/Demangler/Demangler.cs]: Add a new line after and `NumberLen` > `NumberLength`

File: Ryujinx.HLE/HOS/Diagnostics/Demangler/Demangler.cs [hunk 5]
```diff
@@ -0,0 +1,3316 @@
...
+            for (NumberLen = 0; NumberLen < Part.Length; NumberLen++)
+            {
+                if (!char.IsDigit(Part[NumberLen]))
+                {
+                    break;
+                }
+            }
+            Pos += NumberLen;
```
Comment (AcK77) [inline Ryujinx.HLE/HOS/Diagnostics/Demangler/Demangler.cs]: Add a new line before and after

File: Ryujinx.HLE/HOS/Diagnostics/Demangler/Demangler.cs [hunk 6]
```diff
@@ -0,0 +1,3316 @@
...
+                ConsumeIf("n");
+            }
+
+            if (Count() == 0 || !char.IsDigit(Mangled[Pos]))
+            {
+                return null;
+            }
+
+            string Part = Mangled.Substring(Pos);
+            int NumberLen;
```
Comment (AcK77) [inline Ryujinx.HLE/HOS/Diagnostics/Demangler/Demangler.cs]: Add a new line after and NumberLen > NumberLength

File: Ryujinx.HLE/HOS/Diagnostics/Demangler/Demangler.cs [hunk 7]
```diff
@@ -0,0 +1,3316 @@
...
+            for (NumberLen = 0; NumberLen < Part.Length; NumberLen++)
+            {
+                if (!char.IsDigit(Part[NumberLen]))
+                {
+                    break;
+                }
+            }
+            Pos += NumberLen;
```
Comment (AcK77) [inline Ryujinx.HLE/HOS/Diagnostics/Demangler/Demangler.cs]: Add a new line before and after

File: Ryujinx.HLE/HOS/Diagnostics/Demangler/Demangler.cs [hunk 8]
```diff
@@ -0,0 +1,3316 @@
...
+                            return new NameType("operator()");
+                        case 'm':
+                            Pos += 2;
+                            return new NameType("operator,");
+                        case 'o':
+                            Pos += 2;
+                            return new NameType("operator~");
+                        case 'v':
+                            Pos += 2;
+                            bool CanParseTemplateArgsBackup = CanParseTemplateArgs;
```
Comment (AcK77) [inline Ryujinx.HLE/HOS/Diagnostics/Demangler/Demangler.cs]: Maybe you could align some assignment here

File: Ryujinx.HLE/HOS/Diagnostics/Demangler/Demangler.cs [hunk 9]
```diff
@@ -0,0 +1,3316 @@
...
+
+        // <unqualified-name> ::= <operator-name> [<abi-tags> (TODO)]
+        //                    ::= <ctor-dtor-name> (TODO)
+        //                    ::= <source-name>
+        //                    ::= <unnamed-type-name> (TODO)
+        //                    ::= DC <source-name>+ E      # structured binding declaration (TODO)
+        private BaseNode ParseUnqualifiedName(NameParserContext Context)
+        {
+            BaseNode Res = null;
+            char C = Peek();
```
Comment (AcK77) [inline Ryujinx.HLE/HOS/Diagnostics/Demangler/Demangler.cs]: Better name for `C` ? and add a new line before

File: Ryujinx.HLE/HOS/Diagnostics/Demangler/Demangler.cs [hunk 10]
```diff
@@ -0,0 +1,3316 @@
...
+            }
+            else
+            {
+                Res = ParseOperatorName(Context);
+            }
+
+            if (Res != null)
+            {
+                // TODO: ABI Tags
+                //throw new Exception("ABI Tags not implemented");
```
Comment (AcK77) [inline Ryujinx.HLE/HOS/Diagnostics/Demangler/Demangler.cs]: leftover ?

File: Ryujinx.HLE/HOS/Diagnostics/Demangler/Demangler.cs [hunk 11]
```diff
@@ -0,0 +1,3316 @@
...
+            }
+            return Res;
```
Comment (AcK77) [inline Ryujinx.HLE/HOS/Diagnostics/Demangler/Demangler.cs]: Add a new line before

File: Ryujinx.HLE/HOS/Diagnostics/Demangler/Demangler.cs [hunk 12]
```diff
@@ -0,0 +1,3316 @@
...
+        {
+            if (Prev.Type == NodeType.SpecialSubstitution && Prev is SpecialSubstitution)
+            {
+                ((SpecialSubstitution)Prev).SetExtended();
+            }
+
+            if (ConsumeIf("C"))
+            {
+                bool IsInherited = ConsumeIf("I");
+                char C = Peek();
```
Comment (AcK77) [inline Ryujinx.HLE/HOS/Diagnostics/Demangler/Demangler.cs]: Better name for C ? Add a new line before and remove the empty line after

File: Ryujinx.HLE/HOS/Diagnostics/Demangler/Demangler.cs [hunk 13]
```diff
@@ -0,0 +1,3316 @@
...
+
+                if (C != '1' && C != '2' && C != '3')
+                {
+                    return null;
+                }
+
+                Pos++;
```
Comment (AcK77) [inline Ryujinx.HLE/HOS/Diagnostics/Demangler/Demangler.cs]: Add a new line after

File: Ryujinx.HLE/HOS/Diagnostics/Demangler/Demangler.cs [hunk 14]
```diff
@@ -0,0 +1,3316 @@
...
+
+            if (ConsumeIf("D"))
+            {
+                char C = Peek();
+                if (C != '0' && C != '1' && C != '2')
+                {
+                    return null;
+                }
+
+                Pos++;
```
Comment (AcK77) [inline Ryujinx.HLE/HOS/Diagnostics/Demangler/Demangler.cs]: Add a new line after

File: Ryujinx.HLE/HOS/Diagnostics/Demangler/Demangler.cs [hunk 15]
```diff
@@ -0,0 +1,3316 @@
...
+                if (Context != null)
+                {
+                    Context.CtorDtorConversion = true;
+                }
+
+                return new CtorDtorNameType(Prev, true);
+            }
+            return null;
```
Comment (AcK77) [inline Ryujinx.HLE/HOS/Diagnostics/Demangler/Demangler.cs]: Add a new line before

File: Ryujinx.HLE/HOS/Diagnostics/Demangler/Demangler.cs [hunk 16]
```diff
@@ -0,0 +1,3316 @@
...
+        // <function-param> ::= fp <top-level CV-qualifiers> _                                                                                           # L == 0, first parameter
+        //                  ::= fp <top-level CV-qualifiers> <parameter-2 non-negative number> _                                                         # L == 0, second and later parameters
+        //                  ::= fL <L-1 non-negative number> p <top-level CV-qualifiers> _                                                               # L > 0, first parameter
+        //                  ::= fL <L-1 non-negative number> p <top-level CV-qualifiers> <parameter-2 non-negative number> _                             # L > 0, second and later parameters
+        private BaseNode ParseFunctionParameter()
+        {
+            if (ConsumeIf("fp"))
+            {
+                // ignored
+                ParseCVQualifiers();
```
Comment (AcK77) [inline Ryujinx.HLE/HOS/Diagnostics/Demangler/Demangler.cs]: Add a new line after

File: Ryujinx.HLE/HOS/Diagnostics/Demangler/Demangler.cs [hunk 17]
```diff
@@ -0,0 +1,3316 @@
...
+                if (L1Number == null || L1Number.Length == 0)
+                {
+                    return null;
+                }
+
+                if (!ConsumeIf("p"))
+                {
+                    return null;
+                }
+                // ignored
```
Comment (AcK77) [inline Ryujinx.HLE/HOS/Diagnostics/Demangler/Demangler.cs]: Add a new line before

File: Ryujinx.HLE/HOS/Diagnostics/Demangler/Demangler.cs [hunk 18]
```diff
@@ -0,0 +1,3316 @@
...
+                ParseCVQualifiers();
+                if (!ConsumeIf("_"))
+                {
+                    return null;
+                }
+
+                return new FunctionParameter(ParseNumber());
+
+            }
+            return null;
```
Comment (AcK77) [inline Ryujinx.HLE/HOS/Diagnostics/Demangler/Demangler.cs]: Add a new line before

File: Ryujinx.HLE/HOS/Diagnostics/Demangler/Demangler.cs [hunk 19]
```diff
@@ -0,0 +1,3316 @@
...
+        //             ::= fl <binary-operator-name> <expression>
+        //             ::= fr <binary-operator-name> <expression>
+        private BaseNode ParseFoldExpression()
+        {
+            if (!ConsumeIf("f"))
+            {
+                return null;
+            }
+
+            char FoldKind = Peek();
```
Comment (AcK77) [inline Ryujinx.HLE/HOS/Diagnostics/Demangler/Demangler.cs]: Align some var here

File: Ryujinx.HLE/HOS/Diagnostics/Demangler/Demangler.cs [hunk 20]
```diff
@@ -0,0 +1,3316 @@
...
+                case "rs":
+                    OperatorName = ">>";
+                    break;
+                case "rS":
+                    OperatorName = ">>=";
+                    break;
+                default:
+                    return null;
+            }
+            Pos += 2;
```
Comment (AcK77) [inline Ryujinx.HLE/HOS/Diagnostics/Demangler/Demangler.cs]: Add a new line before

File: Ryujinx.HLE/HOS/Diagnostics/Demangler/Demangler.cs [hunk 21]
```diff
@@ -0,0 +1,3316 @@
...
+
+            BaseNode Expression = ParseExpression();
+            BaseNode Initializer = null;
```
Comment (AcK77) [inline Ryujinx.HLE/HOS/Diagnostics/Demangler/Demangler.cs]: Maybe you can move this one after the `if` ?

File: Ryujinx.HLE/HOS/Diagnostics/Demangler/Demangler.cs [hunk 22]
```diff
@@ -0,0 +1,3316 @@
...
+
+            if (HasInitializer)
+            {
+                Initializer = ParseExpression();
+                if (Initializer == null)
+                {
+                    return null;
+                }
+            }
+            if (IsLeftFold && Initializer != null)
```
Comment (AcK77) [inline Ryujinx.HLE/HOS/Diagnostics/Demangler/Demangler.cs]: Add a new line before

File: Ryujinx.HLE/HOS/Diagnostics/Demangler/Demangler.cs [hunk 23]
```diff
@@ -0,0 +1,3316 @@
...
+            {
+                BaseNode Temp = Expression;
```
Comment (AcK77) [inline Ryujinx.HLE/HOS/Diagnostics/Demangler/Demangler.cs]: Align some var here

File: Ryujinx.HLE/HOS/Diagnostics/Demangler/Demangler.cs [hunk 24]
```diff
@@ -0,0 +1,3316 @@
...
+                Expression = Initializer;
+                Initializer = Temp;
+            }
+            return new FoldExpression(IsLeftFold, OperatorName, new PackedTemplateParameterExpansion(Expression), Initializer);
```
Comment (AcK77) [inline Ryujinx.HLE/HOS/Diagnostics/Demangler/Demangler.cs]: Add a new line before

File: Ryujinx.HLE/HOS/Diagnostics/Demangler/Demangler.cs [hunk 25]
```diff
@@ -0,0 +1,3316 @@
...
+        //                ::= cv <type> <expression>                               # type (expression), conversion with one argument
+        //                ::= cv <type> _ <expression>* E                          # type (expr-list), conversion with other than one argument
+        private BaseNode ParseConversionExpression()
+        {
+            if (!ConsumeIf("cv"))
+            {
+                return null;
+            }
+
+            bool CanParseTemplateArgsBackup = CanParseTemplateArgs;
```
Comment (AcK77) [inline Ryujinx.HLE/HOS/Diagnostics/Demangler/Demangler.cs]: Align some var here

File: Ryujinx.HLE/HOS/Diagnostics/Demangler/Demangler.cs [hunk 26]
```diff
@@ -0,0 +1,3316 @@
...
+                        BracedExpressionNode = ParseBracedExpression();
+                        if (BracedExpressionNode == null)
+                        {
+                            return null;
+                        }
+
+                        return new BracedRangeExpression(RangeBeginExpression, RangeEndExpression, BracedExpressionNode);
+                }
+            }
+            return ParseExpression();
```
Comment (AcK77) [inline Ryujinx.HLE/HOS/Diagnostics/Demangler/Demangler.cs]: Add a new line before

File: Ryujinx.HLE/HOS/Diagnostics/Demangler/Demangler.cs [hunk 27]
```diff
@@ -0,0 +1,3316 @@
...
+
+        //               ::= [gs] nw <expression>* _ <type> E                    # new (expr-list) type
+        //               ::= [gs] nw <expression>* _ <type> <initializer>        # new (expr-list) type (init)
+        //               ::= [gs] na <expression>* _ <type> E                    # new[] (expr-list) type
+        //               ::= [gs] na <expression>* _ <type> <initializer>        # new[] (expr-list) type (init)
+        //
+        // <initializer> ::= pi <expression>* E                                  # parenthesized initialization
+        private BaseNode ParseNewExpression()
+        {
+            bool IsGlobal = ConsumeIf("gs");
```
Comment (AcK77) [inline Ryujinx.HLE/HOS/Diagnostics/Demangler/Demangler.cs]: Align some var here

File: Ryujinx.HLE/HOS/Diagnostics/Demangler/Demangler.cs [hunk 28]
```diff
@@ -0,0 +1,3316 @@
...
+            bool IsArray = Peek(1) == 'a';
+
+            if (!ConsumeIf("nw") || !ConsumeIf("na"))
+            {
+                return null;
+            }
+
+            List<BaseNode> Expressions = new List<BaseNode>();
```
Comment (AcK77) [inline Ryujinx.HLE/HOS/Diagnostics/Demangler/Demangler.cs]: Alignment

File: Ryujinx.HLE/HOS/Diagnostics/Demangler/Demangler.cs [hunk 29]
```diff
@@ -0,0 +1,3316 @@
...
+            {
+                BaseNode Expression = ParseExpression();
+                if (Expression == null)
+                {
+                    return null;
+                }
+
+                Expressions.Add(Expression);
+            }
+            BaseNode TypeNode = ParseType();
```
Comment (AcK77) [inline Ryujinx.HLE/HOS/Diagnostics/Demangler/Demangler.cs]: Add a new line before

File: Ryujinx.HLE/HOS/Diagnostics/Demangler/Demangler.cs [hunk 30]
```diff
@@ -0,0 +1,3316 @@
...
+        //              ::= az <expression>                                      # alignof (expression)
+        //              ::= nx <expression>                                      # noexcept (expression)
+        //              ::= <template-param>
+        //              ::= <function-param>
+        //              ::= dt <expression> <unresolved-name>                    # expr.name
+        //              ::= pt <expression> <unresolved-name>                    # expr->name
+        //              ::= ds <expression> <expression>                         # expr.*expr
+        //              ::= sZ <template-param>                                  # sizeof...(T), size of a template parameter pack
+        //              ::= sZ <function-param>                                  # sizeof...(parameter), size of a function parameter pack
+        //              ::= sP <template-arg>* E                                 # sizeof...(T), size of a captured template parameter pack from an alias
```
Comment (AcK77) [inline Ryujinx.HLE/HOS/Diagnostics/Demangler/Demangler.cs]: Add a new line after

Comment (AcK77) [inline Ryujinx.HLE/HOS/Diagnostics/Demangler/Demangler.cs]: Alignment

Comment (AcK77) [inline Ryujinx.HLE/HOS/Diagnostics/Demangler/Demangler.cs]: Add a new line before

Comment (AcK77) [inline Ryujinx.HLE/HOS/Diagnostics/Demangler/Demangler.cs]: Add a new line before

Comment (AcK77) [inline Ryujinx.HLE/HOS/Diagnostics/Demangler/Demangler.cs]: Add a new line before

Comment (AcK77) [inline Ryujinx.HLE/HOS/Diagnostics/Demangler/Demangler.cs]: Add a new line before

Comment (AcK77) [inline Ryujinx.HLE/HOS/Diagnostics/Demangler/Demangler.cs]: Add a new line after

Comment (AcK77) [inline Ryujinx.HLE/HOS/Diagnostics/Demangler/Demangler.cs]: Add a new line before

Comment (AcK77) [inline Ryujinx.HLE/HOS/Diagnostics/Demangler/Demangler.cs]: Add a new line after

Comment (AcK77) [inline Ryujinx.HLE/HOS/Diagnostics/Demangler/Demangler.cs]: nit:
```
List<BaseNode> TemplateParamListTemp = new List<BaseNode>(TemplateParamList);
TemplateParamList = TemplateParamListTemp;

BaseNode TemplateArgument = ParseTemplateArgument();
```

Comment (AcK77) [inline Ryujinx.HLE/HOS/Diagnostics/Demangler/Demangler.cs]: Add new line before

Comment (AcK77) [inline Ryujinx.HLE/HOS/Diagnostics/Demangler/Demangler.cs]: Add new line before

Comment (AcK77) [inline Ryujinx.HLE/HOS/Diagnostics/Demangler/Demangler.cs]: Alignment

Comment (AcK77) [inline Ryujinx.HLE/HOS/Diagnostics/Demangler/Demangler.cs]: Add new line before

Comment (AcK77) [inline Ryujinx.HLE/HOS/Diagnostics/Demangler/Demangler.cs]: Add new line before

Comment (AcK77) [inline Ryujinx.HLE/HOS/Diagnostics/Demangler/Demangler.cs]: Add new line before

Comment (AcK77) [inline Ryujinx.HLE/HOS/Diagnostics/Demangler/Demangler.cs]: Add new line before

Comment (AcK77) [inline Ryujinx.HLE/HOS/Diagnostics/Demangler/Demangler.cs]: Add new line before

Comment (AcK77) [inline Ryujinx.HLE/HOS/Diagnostics/Demangler/Demangler.cs]: Add new line before

Comment (AcK77) [inline Ryujinx.HLE/HOS/Diagnostics/Demangler/Demangler.cs]: Add new line after

Comment (AcK77) [inline Ryujinx.HLE/HOS/Diagnostics/Demangler/Demangler.cs]: Add new line before

Comment (AcK77) [inline Ryujinx.HLE/HOS/Diagnostics/Demangler/Demangler.cs]: Add new line before

Comment (AcK77) [inline Ryujinx.HLE/HOS/Diagnostics/Demangler/Demangler.cs]: Add new line after

Comment (AcK77) [inline Ryujinx.HLE/HOS/Diagnostics/Demangler/Demangler.cs]: Add new line before and maybe a better name for `C`

Comment (AcK77) [inline Ryujinx.HLE/HOS/Diagnostics/Demangler/Demangler.cs]: Add new line before

Comment (AcK77) [inline Ryujinx.HLE/HOS/Diagnostics/Demangler/Demangler.cs]: Add new line before

Comment (AcK77) [inline Ryujinx.HLE/HOS/Diagnostics/Demangler/Demangler.cs]: Add new line before

Comment (AcK77) [inline Ryujinx.HLE/HOS/Diagnostics/Demangler/Demangler.cs]: Add new line before

Comment (AcK77) [inline Ryujinx.HLE/HOS/Diagnostics/Demangler/Demangler.cs]: Add new line before

Comment (AcK77) [inline Ryujinx.HLE/HOS/Diagnostics/Demangler/Demangler.cs]: Add new line before

Comment (AcK77) [inline Ryujinx.HLE/HOS/Diagnostics/Demangler/Demangler.cs]: Add new line before

Comment (AcK77) [inline Ryujinx.HLE/HOS/Diagnostics/Demangler/Demangler.cs]: Add new line before

Comment (AcK77) [inline Ryujinx.HLE/HOS/Diagnostics/Demangler/Demangler.cs]: Add new line after

Comment (AcK77) [inline Ryujinx.HLE/HOS/Diagnostics/Demangler/Demangler.cs]: Add new line before

Comment (AcK77) [inline Ryujinx.HLE/HOS/Diagnostics/Demangler/Demangler.cs]: Add new line before

Comment (AcK77) [inline Ryujinx.HLE/HOS/Diagnostics/Demangler/Demangler.cs]: Align names here

Comment (AcK77) [inline Ryujinx.HLE/HOS/Diagnostics/Demangler/Demangler.cs]: Add new line before

Comment (jduncanator) [reply Ryujinx.HLE/HOS/Diagnostics/Demangler/Demangler.cs]: Why are you suggesting changing the order of these statements around? From the looks of the code, the call to ParseTemplateArgument() is deliberately intended to be placed between the cloning of the template parameter list, and the assignment of it. Changing the ordering will result in completely different behavior. Am I missing something?

Comment (Thog) [reply Ryujinx.HLE/HOS/Diagnostics/Demangler/Demangler.cs]: Yup you are right, this is done on purpose.

Comment (AcK77) [reply Ryujinx.HLE/HOS/Diagnostics/Demangler/Demangler.cs]: Yeah you have right, I have missing something here! Sorry for my mistake!

Comment (jduncanator) [reply Ryujinx.HLE/HOS/Diagnostics/Demangler/Demangler.cs]: No problem! This is why we have code review 😄

**Comment by gdkchan** [comment]
I'm ok with merging this code (mostly, will comment on the few style issues I found) as is (in relation to code style), I don't like being too picky. It seems that you're very close to 100%, do you plan to finish it? If so I think its worth anything. Great job by the way!

File: Ryujinx.HLE/HOS/Diagnostics/Demangler/Ast/ConversionOperatorType.cs [hunk 31]
```diff
@@ -0,0 +1,18 @@
+using System.IO;
+
+namespace Ryujinx.HLE.HOS.Diagnostics.Demangler.Ast
+{
+    public class ConversionOperatorType : ParentNode
+    {
+        public ConversionOperatorType(BaseNode Child) : base(NodeType.ConversionOperatorType, Child)
+        {
+
```
[the remark this answers is not in the document]
Comment (gdkchan) [reply Ryujinx.HLE/HOS/Diagnostics/Demangler/Ast/ConversionOperatorType.cs]: I usually use a single line with ` { }` at the end.

**Comment by Thog** [comment]
The only major missing part for the 100% is lambda and I don't think that is worth the effort of implementing it.

File: Ryujinx.HLE/HOS/Diagnostics/Demangler/Ast/CtorDtorNameType.cs [hunk 32]
```diff
@@ -0,0 +1,23 @@
+using System.IO;
+
+namespace Ryujinx.HLE.HOS.Diagnostics.Demangler.Ast
+{
+    public class CtorDtorNameType : ParentNode
+    {
+        bool IsDestructor;
```
[the remark this answers is not in the document]
Comment (gdkchan) [reply Ryujinx.HLE/HOS/Diagnostics/Demangler/Ast/CtorDtorNameType.cs]: Also `private` prefix like the other fields?

File: Ryujinx.HLE/HOS/Diagnostics/Demangler/Ast/DtorName.cs [hunk 33]
```diff
@@ -0,0 +1,17 @@
+using System.IO;
+
+namespace Ryujinx.HLE.HOS.Diagnostics.Demangler.Ast
+{
+    public class DtorName : ParentNode
+    {
+        public DtorName(BaseNode Name) : base(NodeType.DtOrName, Name)
```
Comment (gdkchan) [inline Ryujinx.HLE/HOS/Diagnostics/Demangler/Ast/DtorName.cs]: Same comment as the other class.

File: Ryujinx.HLE/HOS/Diagnostics/Demangler/Ast/DynamicExceptionSpec.cs [hunk 34]
```diff
@@ -0,0 +1,18 @@
+using System.IO;
+
+namespace Ryujinx.HLE.HOS.Diagnostics.Demangler.Ast
+{
+    public class DynamicExceptionSpec : ParentNode
+    {
+        public DynamicExceptionSpec(BaseNode Child) : base(NodeType.DynamicExceptionSpec, Child)
```
Comment (gdkchan) [inline Ryujinx.HLE/HOS/Diagnostics/Demangler/Ast/DynamicExceptionSpec.cs]: Same comment as the other class.

File: Ryujinx.HLE/HOS/Diagnostics/Demangler/Ast/LiteralOperator.cs [hunk 35]
```diff
@@ -0,0 +1,19 @@
+using System.IO;
+
+namespace Ryujinx.HLE.HOS.Diagnostics.Demangler.Ast
+{
+    public class LiteralOperator : ParentNode
+    {
+        public LiteralOperator(BaseNode Child) : base(NodeType.LiteralOperator, Child)
+        {
+
```
[the remark this answers is not in the document]
Comment (gdkchan) [reply Ryujinx.HLE/HOS/Diagnostics/Demangler/Ast/LiteralOperator.cs]: Same comment as other classes (including `GlobalQualifiedName `), will stop commenting about this issue.

File: Ryujinx.HLE/HOS/Diagnostics/Demangler/Ast/Qualifier.cs [hunk 36]
```diff
@@ -0,0 +1,117 @@
...
+            this.Qualifier = Qualifier;
+        }
+
+        public void PrintQualifier(TextWriter Writer)
+        {
+            if ((Qualifier & CV.Const) != 0)
+            {
+                Writer.Write(" const");
+            }
+            if ((Qualifier & CV.Volatile) != 0)
```
Comment (gdkchan) [inline Ryujinx.HLE/HOS/Diagnostics/Demangler/Ast/Qualifier.cs]: nit: New line between ifs.

File: Ryujinx.HLE/HOS/Diagnostics/Demangler/Ast/Qualifier.cs [hunk 37]
```diff
@@ -0,0 +1,117 @@
...
+            this.Qualifier = Qualifier;
+        }
+
+        public void PrintQualifier(TextWriter Writer)
+        {
+            if ((Qualifier & Reference.LValue) != 0)
+            {
+                Writer.Write("&");
+            }
+            if ((Qualifier & Reference.RValue) != 0)
```
Comment (gdkchan) [inline Ryujinx.HLE/HOS/Diagnostics/Demangler/Ast/Qualifier.cs]: Same as above.

File: Ryujinx.HLE/HOS/Diagnostics/Demangler/Demangler.cs [hunk 38]
```diff
@@ -0,0 +1,3316 @@
...
+                ElementType = ParseType();
+                if (ElementType == null)
+                {
+                    return null;
+                }
+
+                return new ArrayType(ElementType, DimensionExpression);
+            }
+            ElementType = ParseType();
+            if (ElementType == null)
```
Comment (gdkchan) [inline Ryujinx.HLE/HOS/Diagnostics/Demangler/Demangler.cs]: Add a new life before.

File: Ryujinx.HLE/HOS/Diagnostics/Demangler/Demangler.cs [hunk 39]
```diff
@@ -0,0 +1,3316 @@
...
+            }
+
+            BaseNode Res = null;
+            switch (Peek())
+            {
+                case 'r':
+                case 'V':
+                case 'K':
+                    int TypePos = 0;
+                    if (Peek(TypePos) == 'r')
```
Comment (gdkchan) [inline Ryujinx.HLE/HOS/Diagnostics/Demangler/Demangler.cs]: Add a new line before.

File: Ryujinx.HLE/HOS/Diagnostics/Demangler/Demangler.cs [hunk 40]
```diff
@@ -0,0 +1,3316 @@
...
+                            return null;
+                        }
+
+                        Res = new NameTypeWithTemplateArguments(Res, TemplateArguments);
+                    }
+                    break;
+                case 'P':
+                    Pos++;
+                    Res = ParseType(Context);
+                    if (Res == null)
```
Comment (gdkchan) [inline Ryujinx.HLE/HOS/Diagnostics/Demangler/Demangler.cs]: Add a new line before.

File: Ryujinx.HLE/HOS/Diagnostics/Demangler/Demangler.cs [hunk 41]
```diff
@@ -0,0 +1,3316 @@
...
+                    {
+                        return null;
+                    }
+
+                    Res = new PointerType(Res);
+                    break;
+                case 'R':
+                    Pos++;
+                    Res = ParseType(Context);
+                    if (Res == null)
```
Comment (gdkchan) [inline Ryujinx.HLE/HOS/Diagnostics/Demangler/Demangler.cs]: nit: New line before (will stop commenting about this type of issue).

**Comment by Thog** [comment]
Should be good now

File: Ryujinx.HLE/HOS/Diagnostics/Demangler/Demangler.cs [hunk 42]
```diff
@@ -0,0 +1,3367 @@
...
+                Context.FinishWithTemplateArguments = false;
+            }
+
+            return Result;
+        }
+
+        private int ParsePositiveNumber()
+        {
+            string Part      = Mangled.Substring(Position);
+            int NumberLength = 0;
```
Comment (AcK77) [inline Ryujinx.HLE/HOS/Diagnostics/Demangler/Demangler.cs]: Align var names

File: Ryujinx.HLE/HOS/Diagnostics/Demangler/Demangler.cs [hunk 43]
```diff
@@ -0,0 +1,3367 @@
...
+                ConsumeIf("n");
+            }
+
+            if (Count() == 0 || !char.IsDigit(Mangled[Position]))
+            {
+                return null;
+            }
+
+            string Part      = Mangled.Substring(Position);
+            int NumberLength = 0;
```
Comment (AcK77) [inline Ryujinx.HLE/HOS/Diagnostics/Demangler/Demangler.cs]: Align var names

File: Ryujinx.HLE/HOS/Diagnostics/Demangler/Demangler.cs [hunk 44]
```diff
@@ -0,0 +1,3367 @@
...
+        //              ::= [gs] da <expression>                                 # delete[] expression
+        //              ::= dc <type> <expression>                               # dynamic_cast<type> (expression)
+        //              ::= sc <type> <expression>                               # static_cast<type> (expression)
+        //              ::= cc <type> <expression>                               # const_cast<type> (expression)
+        //              ::= rc <type> <expression>                               # reinterpret_cast<type> (expression)
+        //              ::= ti <type>                                            # typeid (type)
+        //              ::= te <expression>                                      # typeid (expression)
+        //              ::= st <type>                                            # sizeof (type)
+        //              ::= sz <expression>                                      # sizeof (expression)
+        //              ::= at <type>
```
Comment (AcK77) [inline Ryujinx.HLE/HOS/Diagnostics/Demangler/Demangler.cs]: Align var names and assignment

Comment (AcK77) [inline Ryujinx.HLE/HOS/Diagnostics/Demangler/Demangler.cs]: Add new line before

Comment (AcK77) [inline Ryujinx.HLE/HOS/Diagnostics/Demangler/Demangler.cs]: Add new line before

Comment (AcK77) [inline Ryujinx.HLE/HOS/Diagnostics/Demangler/Demangler.cs]: Align var names and assignment

Comment (AcK77) [inline Ryujinx.HLE/HOS/Diagnostics/Demangler/Demangler.cs]: Add new line before

Comment (AcK77) [inline Ryujinx.HLE/HOS/Diagnostics/Demangler/Demangler.cs]: Add new line before

Comment (AcK77) [inline Ryujinx.HLE/HOS/Diagnostics/Demangler/Demangler.cs]: Add new line before

Comment (AcK77) [inline Ryujinx.HLE/HOS/Diagnostics/Demangler/Demangler.cs]: Add new line before

Comment (AcK77) [inline Ryujinx.HLE/HOS/Diagnostics/Demangler/Demangler.cs]: Add new line before

Comment (AcK77) [inline Ryujinx.HLE/HOS/Diagnostics/Demangler/Demangler.cs]: Add new line before

Comment (AcK77) [inline Ryujinx.HLE/HOS/Diagnostics/Demangler/Demangler.cs]: Add new line before

**Comment by AcK77** [comment]
lgtm! Thanks!

**Pull request merged**